### PR TITLE
Add fuzzer for HCLK_IOI3 PIPs

### PR DESCRIPTION
**Pull request by tmichalak**
This PR solves #891 
It's WIP as it is still missing some PIPs that drive the BUFIO and BUFR from ILOGIC clock or through the IOI tile.
Currently it solves 149 PIPs out of 196, and the solved ones include all PIPs that are used by Litex on Arty.
The current approach generates the netlist in a partly random fashion and uses Vivado only for P&R and bitstream generation. The fuzzer could be optimized by adding a todo list based rerouting script to Vivado.

### CODE REVIEW

File: fuzzers/047-hclk-ioi-pips/Makefile [hunk 1]
```diff
@@ -0,0 +1,45 @@
+export FUZDIR=$(shell pwd)
+PIP_TYPE?=hclk_ioi3
+PIPLIST_TCL=$(FUZDIR)/hclk_ioi3_pip_list.tcl
+TODO_RE=".*"
```
Comment (litghost) [inline fuzzers/047-hclk-ioi-pips/Makefile]: Change this TODO_RE to filter only include pips we are looking for right now, and we can land this PR.

File: fuzzers/047-hclk-ioi-pips/Makefile [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+MAKETODO_FLAGS= --sides "" --pip-type ${PIP_TYPE} --seg-type ${PIP_TYPE} --re $(TODO_RE)
+N = 50
+
+# These PIPs all appear to be either a 1 bit solutions.
```
Comment (litghost) [inline fuzzers/047-hclk-ioi-pips/Makefile]: Comment is out of date

File: fuzzers/047-hclk-ioi-pips/generate.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            tile_type, dst, src = l.strip().split('.')
+            if tile_type not in pipdata:
+                pipdata[tile_type] = []
+                tile_ports[tile_type] = set()
+
+            pipdata[tile_type].append((src, dst))
+            tile_ports[tile_type].add(src)
+            tile_ports[tile_type].add(dst)
+
+    with open(os.path.join(os.getenv('FUZDIR'), '..', 'piplist', 'build',
```
Comment (litghost) [inline fuzzers/047-hclk-ioi-pips/generate.py]: No need to have the same piplist twice

File: fuzzers/047-hclk-ioi-pips/generate.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        for src, dst in pipdata[tile_type]:
+            if (src, dst) in ignpip:
+                pass
+            elif (src, dst) in pips:
+                segmk.add_tile_tag(tile, "%s.%s" % (dst, src), 1)
+            elif dst not in tiledata[tile]["dsts"]:
+                segmk.add_tile_tag(tile, "%s.%s" % (dst, src), 0)
+
+        #for port in tile_ports[tile_type]:
```
Comment (litghost) [inline fuzzers/047-hclk-ioi-pips/generate.py]: Remove comments?

File: fuzzers/047-hclk-ioi-pips/output_cmt.tcl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+create_project -force -part $::env(XRAY_PART) design design
```
Comment (litghost) [inline fuzzers/047-hclk-ioi-pips/output_cmt.tcl]: Should we rename this file to something more descriptive?

Comment (litghost) [reply fuzzers/047-hclk-ioi-pips/output_cmt.tcl]: Sorry this comment got attached to the wrong file.

**Comment by tmichalak** [comment]
Red CI due to timeout. Without targeted rerouting we want be able to get all of the PIPs solved as they appear to rarely in the specimen

File: fuzzers/047-hclk-ioi-pips/Makefile (hunk 2 above)
Comment (tmichalak) [reply fuzzers/047-hclk-ioi-pips/Makefile]: Removed

File: fuzzers/047-hclk-ioi-pips/generate.py (hunk 3 above)
Comment (tmichalak) [reply fuzzers/047-hclk-ioi-pips/generate.py]: Removed the duplicate entry

File: fuzzers/047-hclk-ioi-pips/generate.py (hunk 4 above)
Comment (tmichalak) [reply fuzzers/047-hclk-ioi-pips/generate.py]: Removed

File: fuzzers/047-hclk-ioi-pips/Makefile (hunk 1 above)
Comment (tmichalak) [reply fuzzers/047-hclk-ioi-pips/Makefile]: Added an EXCLUDE_RE and also min-progress to be able to land this PR.
The bits used by Artix will be solved by the time we get to the min progress trigger.

**Pull request merged**